fe/fe-core/src/main/java/com/starrocks/metric/ResourceGroupMetricMgr.java

```diff
@@ -0,0 +1,232 @@
+                    RESOURCE_GROUP_QUERY_ERR_COUNTER_MAP.put(resourceGroupName, metric);
+                    MetricRepo.addMetric(metric);
+                    LOG.info("Add {} metric, resource group name is {}", QUERY_RESOURCE_GROUP_ERR, resourceGroupName);
+                }
+            }
+        }
+        RESOURCE_GROUP_QUERY_ERR_COUNTER_MAP.get(resourceGroupName).increase(num);
+    }
+
+    public static void updateQueryLatency(List<QueryDetail> queryList) {
+        Map<String, List<Long>> latencyMap = new HashMap<>();
+        Map<String, Long> latencySumMap = new HashMap<>();
+        for (QueryDetail queryDetail : queryList) {
+            String workGroupName = queryDetail.getWorkGroupName();
+            if (queryDetail.isQuery()
+                    && queryDetail.getState() == QueryDetail.QueryMemState.FINISHED
+                    && workGroupName != null
+                    && !workGroupName.isEmpty()) {
+                if (!latencyMap.containsKey(workGroupName)) {
+                    latencyMap.put(workGroupName, new ArrayList<>());
+                    latencySumMap.put(workGroupName, 0L);
+                }
+                latencyMap.get(workGroupName).add(queryDetail.getLatency());
+                latencySumMap.put(workGroupName, latencySumMap.get(workGroupName) + queryDetail.getLatency());
+            }
+        }
+        for (String resourceGroupName : latencyMap.keySet()) {
+            List<Long> latencyList = latencyMap.get(resourceGroupName);
+            Long latencySum = latencySumMap.get(resourceGroupName);
+
+            if (!RESOURCE_GROUP_QUERY_LATENCY_MAP.containsKey(resourceGroupName)) {
+                createQueryResourceGroupLatency(resourceGroupName);
+            }
+            List<GaugeMetricImpl> metricList = RESOURCE_GROUP_QUERY_LATENCY_MAP.get(resourceGroupName);
+            if (latencyList.size() > 0) {
+                metricList.get(0).setValue(latencySum / latencyList.size());
+
+                latencyList.sort(Comparator.naturalOrder());
+
+                int index = (int) Math.round((latencyList.size() - 1) * 0.5);
+                metricList.get(1).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.75);
+                metricList.get(2).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.90);
+                metricList.get(3).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.95);
+                metricList.get(4).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.99);
+                metricList.get(5).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.999);
+                metricList.get(6).setValue((double) latencyList.get(index));
```

Comment:
Suggest to use a for-loop with a `List {0.5, 0.75, ..}`.

Reply:
Yes, but not necessary, in this case, it may be better to write each metrics clearly

Reply:
IMO, using a `List {0.5, 0.75, ..}` may be more clear.

Reply:
@only2yangcao  suggest to use for-loop,   this code is plain and straightforward, unfolding a for-loop like this make the code hard to maintainment, so keeping the code simple and sucinct is very import and is a good programming habit.

Reply:
OK i will refact code